### ISSUE TITLE
'NoneType' object has no attribute 'x'

### ISSUE THREAD

**Issue by onurays** [Pithikos/python-rectangles#1]
print Rect(577, 184, 62, 87).distance_to_rect(Rect(628, 179, 3, 3))

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "geometry.py", line 194, in distance_to_rect
    distance_between_edge_and_point(edge1, edge2[0]),
  File "geometry.py", line 242, in distance_between_edge_and_point
    if point_faces_edge(edge, point):
  File "geometry.py", line 232, in point_faces_edge
    a=edge[0].distance_to_point(edge[1])
  File "geometry.py", line 29, in distance_to_point
    return sqrt((self.x-p.x)**2+(self.y-p.y)**2)
AttributeError: 'NoneType' object has no attribute 'x'

I think it is related to overlaps_with function. Maybe we should check them in both way like this:

if self.overlaps_with(rect) or rect.overlaps_with(self):
			return 0

**Comment by onurays** [comment]
To avoid the exception, I added this control to distance_to_rect before calculating distances:

if edge1[0] is None or edge1[1] is None or edge2[0] is None or edge2[1] is None:
			return 0

Do you have better suggestions?

**Comment by Pithikos** [comment]
I think the problem is an edge case in `distance_to_rect` at step 3.

The way the distance between rectangles is calculated is by at a point drawing a line between the centers of the two rectangles. At step 3 we try to find the two segments (one for rectangle 1 and one for rectangle 2) that intersect the line we just drew. Now with the current implementation there is a case where one such edge can't be found.

A simplified version that showcases this:

```python
r1 = Rect(0, 5, 50, 50)
r2 = Rect(35, 0, 1, 1)
print(r1.distance_to_rect(r2))
```

**Comment by Pithikos** [comment]
This has now been fixed @39aa41bca1083e0fe5575ed626bd8978b32e4a47

The problem was indeed the calculation when trying to determine edges that intersect a line. The new approach is more generic and should thus not fail.

Feel free to reopen the issue if you think that this doesn't entirely solve it.

**Comment by onurays** [comment]
Thanks for the quick help. However edge2 is still None. Here is the log:

( 608, 198)              ( 623, 198)
      .-----------------------.
      |                       |
      |                  31.0 |
      |         15.0          |
      '-----------------------'
( 608, 229)              ( 623, 229)
( 552, 163)              ( 639, 163)
      .-----------------------.
      |                       |
      |                  93.0 |
      |         87.0          |
      '-----------------------'
( 552, 256)              ( 639, 256)

Traceback (most recent call last):
 ...
  File "geometry.py", line 191, in distance_to_rect
    assert edge2
AssertionError

**Comment by Pithikos** [comment]
I don't know if it's related but after fix 3dcee1786b876117674c0a1371b0a0f1d5fc313f it seems to work fine.

I tested for the last rectangles you mentioned

```
In [45]: Rect(608, 198, 15, 31).distance_to_rect(Rect(552, 163, 87, 93))
Out[45]: 0
```
